gpflow/monitor.py

```diff
@@ -0,0 +1,245 @@
+            Can be a nested. E.g. ./logs/my_run/
+        :param period: defines how often to run the task; it will execute every `period`th step.
+            For large values of `period` the task will be less frequently run.
+        """
+        super().__init__(period)
+        self.file_writer = tf.summary.create_file_writer(log_dir)
+
+    def __call__(self, step, **kwargs):
+        with self.file_writer.as_default():
+            super().__call__(step, **kwargs)
+        self.file_writer.flush()
+
+
+class ModelToTensorBoard(ToTensorBoard):
+    """
+    Monitoring task that creates a sensible TensorBoard for a model.
+
+    By default, it writes to the TensorBoard all model parameters that are scalars,
+    for parameter arrays (e.g. kernel.lengthscales) the values are sent to the TensorBoard
+    if the array has no more than `max_size` (default: 3) elements.
+    This behaviour can be adjusted using `max_size`.
+    """
+
+    def __init__(
+        self,
+        log_dir: str,
+        model: BayesianModel,
+        max_size: int = 3,
+        period: int = 1,
+        extra_keywords: List[str] = [],
+    ):
+        """
+        :param log_dir: directory in which to store the tensorboard files.
+            Can be a nested: for example, './logs/my_run/'.
+        :param model: model to be monitord.
+        :param max_size: maximum size of arrays (incl.) to store each
+            element of the array independently as a scalar in the TensorBoard.
+        :param period: defines how often to run the task; it will execute every `period`th step.
+            For large values of `period` the task will be less frequently run.
+        :param extra_keywords: pass additional keywords to be monitored.
+            If the parameter's name includes any of the extra keywords specified it
+            will be monitored as well. By default, parameters that match the `kernel` or
+            `likelihood` keyword are monitored.
+        """
+        super().__init__(log_dir, period)
+        self.model = model
+        self.max_size = max_size
+        self.keywords_to_monitor = ["kernel", "likelihood"] + extra_keywords
+
+    def run(self, **unused_kwargs):
+        for name, parameter in parameter_dict(self.model).items():
+            # check if the parameter name matches any of the specified keywords
+            if any(keyword in name for keyword in self.keywords_to_monitor):
+                name = name.lstrip(".")  # keys are prepended with a '.', which we strip
+                self._summarize_parameter(name, parameter)
+
+    def _summarize_parameter(self, name: str, value: Union[float, np.ndarray]):
+        """
+        :param name: identifier used in tensorboard
+        :param value: value to be stored in tensorboard
+        """
+        value = tf.reshape(value, (-1,))
+        for i in tf.range(tf.minimum(tf.size(value), self.max_size)):
+            tf.summary.scalar(f"{name}[{i}]", value[i], step=self.current_step)
```

Comment:
does this mean all actually-scalar parameters (e.g. `variance`) would still end up in the tensorboard as `variance[0]` rather than just `variance`? would it be possible to avoid that and only add the index if there is more than one?

Reply:
done. makes it neater indeed.